File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.py
class Solution:  
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        status = {}
        safe_nodes = []
        def is_safe(i):
            if i in status:
                return status[i]
            else:
                status[i] = False
                for nei in graph[i]:
                    if  not is_safe(nei):
                        return status[i]
                status[i] = True
                return status[i]
        for i in range(len(graph)):
            if is_safe(i):
                safe_nodes.append(i)
        return safe_nodes
            
        

File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.cpp

class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        std::map<int, bool> is_safe;
        std::vector<int> safe_nodes; 
        int n = graph.size();
        std::function<bool(int)> check_safety = [&is_safe, graph, &check_safety](int i) -> bool {
            bool res = false;
            if (is_safe.count(i) > 0){
            res = is_safe[i];
            return res;
            }
            is_safe[i] = false;
            for(int nei : graph[i]){
                if (!check_safety(nei)){
                    res = false;
                return res; 
                }
            }
            is_safe[i] = true;
            res = true;
            return res;
        };
        for(int i=0; i<n; i++){
            if (check_safety(i)) {safe_nodes.insert(safe_nodes.end(), i);}
        }
        return safe_nodes;
    }
};